File: samples/PythonContainerApp/AppHost/app_host.py
from pathlib import Path
from urllib.parse import urlparse
import os
import sys

from cloudshell.launcher import CloudShellDistributedApplication, LauncherOptions


app_host_dir = Path(__file__).resolve().parent
sample_dir = app_host_dir.parent
repo_root = sample_dir.parent.parent

app = (
    CloudShellDistributedApplication.create_builder(
        "python-container-app",
        sys.argv[1:],
        base_path=app_host_dir,
    )
    .with_metadata("cloudshell.source", "python")
    .with_metadata("cloudshell.sample", "PythonContainerApp")
)

python_app_path = app.resolve_path("..", "App")

settings_endpoint = os.getenv("CLOUDSHELL_SETTINGS_ENDPOINT", "http://localhost:5111")
secrets_endpoint = os.getenv("CLOUDSHELL_SECRETS_ENDPOINT", "http://localhost:6111")
app_endpoint = urlparse(os.getenv("CLOUDSHELL_APP_ENDPOINT", "http://localhost:5190"))


def define_resources(resources):
    network = resources.default_network()
    settings = (
        resources.add_configuration_store("python-container-settings")
        .with_display_name("Python Container Settings")
        .with_endpoint(settings_endpoint)
        .with_seed(
            lambda seed: seed.setting(
                "Sample--Message",
                "Hello from Python container app configuration",
            )
        )
    )
    secrets = (
        resources.add_secrets_vault("python-container-secrets")
        .with_display_name("Python Container Secrets")
        .with_endpoint(secrets_endpoint)
        .with_seed(
            lambda seed: seed.secret(
                "Sample--ApiKey",
                "python-container-secret",
                "v1",
            )
        )
    )

    api = (
        resources.add_python_app("python-container-api", python_app_path)
        .with_display_name("Python Container API")
        .with_service_discovery()
        .with_reference(settings)
        .with_reference(secrets)
        .depends_on(settings)
        .depends_on(secrets)
        .with_environment_variable("PORT", "8080")
        .with_environment_variable("OTEL_SERVICE_NAME", "python-container-api")
        .with_http_endpoint(
            host=app_endpoint.hostname,
            port=app_endpoint.port,
            target_port=8080,
            network=network,
        )
        .with_http_health_check("/healthz", endpoint_name="http")
        .with_http_liveness_check("/alive", endpoint_name="http")
        .as_container_app(
            tag="dev",
            build_context=repo_root,
            dockerfile="samples/PythonContainerApp/App/Dockerfile",
            replicas=2,
        )
        .require_identity(name="python-container-api")
        .provision_identity_on_startup()
    )
    settings.allow_resource_identity(
        api,
        "CloudShell.Configuration/stores/settings/read/action",
        identity_name="python-container-api",
    )
    secrets.allow_resource_identity(
        api,
        "CloudShell.Secrets/vaults/secrets/read/action",
        identity_name="python-container-api",
    )


app.define_resources(define_resources)

defaults = LauncherOptions(
    cli_project=os.getenv(
        "CLOUDSHELL_CLI_PROJECT",
        str(repo_root / "CloudShell.Cli" / "CloudShell.Cli.csproj"),
    ),
    template_path=os.getenv(
        "CLOUDSHELL_TEMPLATE_PATH",
        str(sample_dir / ".cloudshell" / "resources.json"),
    ),
    control_plane_url=os.getenv(
        "CLOUDSHELL_CONTROL_PLANE_URL",
        "http://127.0.0.1:5110",
    ),
    state_dir=os.getenv("CLOUDSHELL_STATE_DIR", str(sample_dir / ".cloudshell")),
    data_dir=os.getenv(
        "CLOUDSHELL_DATA_DIR",
        str(sample_dir / ".cloudshell" / "data"),
    ),
    host_project=os.getenv(
        "CLOUDSHELL_HOST_PROJECT",
        str(
            repo_root
            / "CloudShell.LocalDevelopmentHost"
            / "CloudShell.LocalDevelopmentHost.csproj"
        ),
    ),
    host_url=os.getenv(
        "CLOUDSHELL_CONTROL_PLANE_URL",
        "http://127.0.0.1:5110",
    ),
)

raise SystemExit(app.run(options=defaults))

File: samples/TypeScriptContainerApp/src/app-host.ts
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { cloudshell } from "@cloudshell/local-development";

const sampleRoot = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const repoRoot = resolve(sampleRoot, "..", "..");
const appRoot = resolve(sampleRoot, "App");
const cliProject = process.env.CLOUDSHELL_CLI_PROJECT ??
  resolve(repoRoot, "CloudShell.Cli", "CloudShell.Cli.csproj");
const hostProject = process.env.CLOUDSHELL_HOST_PROJECT ??
  resolve(repoRoot, "CloudShell.LocalDevelopmentHost", "CloudShell.LocalDevelopmentHost.csproj");
const stateDir = process.env.CLOUDSHELL_STATE_DIR ??
  resolve(sampleRoot, ".cloudshell");
const dataDir = readArgumentValue("--data-dir") ??
  process.env.CLOUDSHELL_DATA_DIR ??
  resolve(stateDir, "data");
const templatePath = process.env.CLOUDSHELL_TEMPLATE_PATH ??
  resolve(stateDir, "resources.json");
const controlPlaneUrl = process.env.CLOUDSHELL_CONTROL_PLANE_URL ??
  "http://127.0.0.1:5114";
const settingsEndpoint = process.env.CLOUDSHELL_SETTINGS_ENDPOINT ??
  "http://localhost:5115";
const secretsEndpoint = process.env.CLOUDSHELL_SECRETS_ENDPOINT ??
  "http://localhost:6115";
const appEndpoint = new URL(process.env.CLOUDSHELL_APP_ENDPOINT ??
  "http://localhost:5192");

const app = cloudshell("typescript-container-app", {
  environmentId: "local",
  metadata: {
    "cloudshell.source": "typescript",
    "cloudshell.sample": "TypeScriptContainerApp"
  }
});

const network = app.getDefaultNetwork();

const settings = app
  .addConfigurationStore("typescript-container-settings")
  .withDisplayName("TypeScript Container Settings")
  .withEndpoint(settingsEndpoint)
  .withSeed(seed => seed
    .setting("Sample--Message", "Hello from TypeScript container app configuration")
    .setting("Sample--Mode", "container"));

const secrets = app
  .addSecretsVault("typescript-container-secrets")
  .withDisplayName("TypeScript Container Secrets")
  .withEndpoint(secretsEndpoint)
  .withSeed(seed => seed.secret("Sample--ApiKey", "typescript-container-secret", "v1"));

const api = app
  .addJavaScriptApp("typescript-container-api", appRoot)
  .withDisplayName("TypeScript Container API")
  .withPackageManager("npm")
  .withScript("start")
  .withServiceDiscovery()
  .withReference(settings)
  .withReference(secrets)
  .dependsOn(settings)
  .dependsOn(secrets)
  .withEnvironmentVariable("PORT", "8080")
  .withEnvironmentVariable("OTEL_SERVICE_NAME", "typescript-container-api")
  .withHttpEndpoint({
    host: appEndpoint.hostname,
    port: appEndpoint.port ? Number(appEndpoint.port) : undefined,
    targetPort: 8080,
    network
  })
  .withHttpHealthCheck("/healthz", { endpointName: "http" })
  .withHttpLivenessCheck("/alive", { endpointName: "http" })
  .asContainerApp({
    tag: "dev",
    buildContext: repoRoot,
    dockerfile: "samples/TypeScriptContainerApp/App/Dockerfile",
    replicas: 2
  })
  .requireIdentity({ name: "typescript-container-api" })
  .provisionIdentityOnStartup();

settings.allowResourceIdentity(
  api,
  "CloudShell.Configuration/stores/settings/read/action",
  { identityName: "typescript-container-api" });
secrets.allowResourceIdentity(
  api,
  "CloudShell.Secrets/vaults/secrets/read/action",
  { identityName: "typescript-container-api" });

if (process.argv.includes("--apply") || process.argv.includes("--start") || process.argv.includes("--run")) {
  const options = {
    cliProject,
    hostProject,
    templatePath,
    noBuild: process.argv.includes("--no-build"),
    controlPlaneUrl,
    url: controlPlaneUrl,
    stateDir,
    dataDir,
    cwd: repoRoot
  };
  const result = process.argv.includes("--run")
    ? await app.run(options)
    : process.argv.includes("--start")
      ? await app.start(options)
      : await app.apply(options);

  process.exitCode = result.exitCode;
} else {
  process.stdout.write(app.toJson());
}

function readArgumentValue(name: string): string | undefined {
  const index = process.argv.findIndex(argument =>
    argument.toLowerCase() === name.toLowerCase());
  return index >= 0 ? process.argv[index + 1] : undefined;
}
